src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -780,6 +781,10 @@ void PSParallelCompact::fill_dense_prefix_end(SpaceId id) {
   // filler obj will extend to next region.
 
   // Note: If min-fill-size decreases to 1, this whole method becomes redundant.
+  if (UseCompactObjectHeaders) {
+    // The gap is always equal to min-fill-size, so nothing to do.
+    return;
+  }
```

Comment:
Reading the comment, it is not obvious that this is correct if you set MinObjectAlignment to something larger than the default value:
```
void PSParallelCompact::fill_dense_prefix_end(SpaceId id) {
  // Comparing two sizes to decide if filling is required:
  //
  // The size of the filler (min-obj-size) is 2 heap words with the default
  // MinObjAlignment, since both markword and klass take 1 heap word.
  //
  // The size of the gap (if any) right before dense-prefix-end is
  // MinObjAlignment.
  //
  // Need to fill in the gap only if it's smaller than min-obj-size, and the
  // filler obj will extend to next region.

  // Note: If min-fill-size decreases to 1, this whole method becomes redundant.
  if (UseCompactObjectHeaders) {
    // The gap is always equal to min-fill-size, so nothing to do.
    return;
  }
  assert(CollectedHeap::min_fill_size() >= 2, "inv");
```

Reply:
Style note: The added code is inserted between a comment and the code that the comment refers to. It would be nice to tidy this up.

Reply:
Did you figure out if the code above is correct w.r.t. `MinObjectAlignment=16`?

Reply:
When MinObjectAlignment=16, then that method does nothing anyway:

```
  if (MinObjAlignment > 1) {
    return;
  }

```

I think what it really means to say is
```
  if (MinObjAlignment >=  CollectedHeap::min_fill_size()) {
    return;
  }

```

That's also what the comment says: "The size of the gap (if any) right before dense-prefix-end is MinObjAlignment. Need to fill in the gap only if it's smaller than min-obj-size, and the filler obj will extend to next region."

If I interpret that correctly, we need to deal with the situation only when MinObjAlignment < min_fill_size, because the filler object would extend to the next region, and we need to adjust the next region and mark-bitmap for that extra word. @albertnetymk might want to confirm.

I'll move the if (UCOH) block down a little bit to right before if (MinObjAlignment) block.

Reply:
After re-reading this again I agree with what you're writing. If you make the change to use:
```
if (MinObjAlignment >= CollectedHeap::min_fill_size()) {
 return;
 }
```

do you even have to check for UCOH in this function?

I also wonder if you could tweak the comment now that this is not true when UCOH is on:
```
// The size of the filler (min-obj-size) is 2 heap words with the default
// MinObjAlignment, since both markword and klass take 1 heap word.
```

Reply:
I took UCOH into account when this code was written -- the current version of PR would fail the following assert. 

```
  // Note: If min-fill-size decreases to 1, this whole method becomes redundant.
  assert(CollectedHeap::min_fill_size() >= 2, "inv");
```

The least intrusive way, IMO, is to put `if (UCOH) { return; }` right before `// Note: ...`, kind of like what Roman originally put it. I believe the advantage of this style is that when UCOH before always-true, it's obvious this whole method essentially becomes `return`and can be removed right away.

Reply:
It's still correct when UCOH is disabled -- therefore, the UCOH check can be placed at the start without changing any existing logic. (The "rest" of this method assumes min-fill-size is 2, `assert(CollectedHeap::min_fill_size() == 2, "inv")`.)

In this PR, since this method doesn't access UCOH, it can be easily forgotten to update this method when the UCOH flag is removed eventually -- it's not obvious to me that `MinObjAlignment >= checked_cast<int>(CollectedHeap::min_fill_size())` is related to (or can be affected by) `UCOH` at first glance.

(I slightly prefer having a `if (UCOH)` inside this method, but considering this method will be nuked in the long run, any short-time decision is fine by me, assuming the failing assert is fixed.)

Reply:
I added an assert(!UCOH) in the implementation so that we don't forget it once the UCOH flag becomes obsolete.